### ISSUE TITLE
Question about F

### ISSUE THREAD

**Issue by FogSue** [Chaos96/fourierft#4]
Thank you very much for providing the open-source code. 
I have a question about the initialization of F. 
F, as a sparse matrix with a shape of (in_dim, out_dim), its non-zero indices are determined by E, where the values of E range from [0, in_dim*out_dim). This seems like it could cause index out-of-bounds errors. 
![2](https://github.com/Chaos96/fourierft/assets/79622142/f1f137e8-8144-4366-a051-10f99d59ff57)

Could you please tell me what details I might be missing?

**Comment by zqgao22** [comment]
> Thank you very much for providing the open-source code. I have a question about the initialization of F. F, as a sparse matrix with a shape of (in_dim, out_dim), its non-zero indices are determined by E, where the values of E range from [0, in_dim*out_dim). This seems like it could cause index out-of-bounds errors. ![2](https://private-user-images.githubusercontent.com/79622142/333889052-f1f137e8-8144-4366-a051-10f99d59ff57.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTY4MTQyNjAsIm5iZiI6MTcxNjgxMzk2MCwicGF0aCI6Ii83OTYyMjE0Mi8zMzM4ODkwNTItZjFmMTM3ZTgtODE0NC00MzY2LWEwNTEtMTBmOTlkNTlmZjU3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTI3VDEyNDYwMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTI2ZDVlNDQ3ODE0MTE2ZDUzMmE0NTMzMjFiMjAwMGIyMjk1Y2Y2NDg4MDRjMmRlYTU5Yzg0YWQ4MmRhOGExMzMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.9oNj4g0JD9vAR5sTThwf7RVszdYa9FHabc1gDBj2k0U)
> 
> Could you please tell me what details I might be missing?

Hello, thank you for pointing this out. It was our mistake. The correct initialization for $E$ here should be `E = torch.randperm(d1 * d2)[:n]` and `E = torch.stack([E // d2, E % d2], dim=0)`. Their sizes are `torch.Size([n])` and `torch.Size([2, n])` respectively. **We will correct this in the code and pseudocode in the paper.** There was no error in the experiments implemented because $d_1$ is equal to $d_2$ for the transformer's Q and V matrices. 

Best,
Ziqi

**Comment by FogSue** [comment]
Thank you very much for your response, I got you now!
